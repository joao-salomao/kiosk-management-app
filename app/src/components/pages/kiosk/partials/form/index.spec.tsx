import renderer from 'react-test-renderer';
import { fireEvent, screen, render, waitFor } from '@testing-library/react';
import { Form } from ".";

it('call the onSubmit handler when all fields are valid', async () => {
  const onSubmitMock = jest.fn();

  render(<Form isSubmitting={false} onSubmit={onSubmitMock} />)

  fireEvent.change(
    screen.getByLabelText(/serial key:/i),
    { target: { value: 'Some serial key' } }
  );

  fireEvent.change(
    screen.getByLabelText(/description:/i),
    { target: { value: 'Some description' } }
  );

  fireEvent.change(
    screen.getByLabelText(/store opens at:/i),
    { target: { value: '10:00' } }
  );

  fireEvent.change(
    screen.getByLabelText(/store closes at:/i),
    { target: { value: '22:00' } }
  );

  fireEvent.click(screen.getByRole('button'));

  await waitFor(() => {
    expect(onSubmitMock).toHaveBeenCalled();
  });
});

describe('when is submitting the form', () => {
  it('disable the submit button', () => {
    render(<Form isSubmitting={true} onSubmit={jest.fn()} />)

    const button = screen.getByRole('button');

    expect(button).toBeDisabled();
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Form isSubmitting={true} onSubmit={jest.fn()} />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe('when is not submitting the form', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Form isSubmitting={false} onSubmit={jest.fn()} />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

it('show the required validation text when the fields values are invalid', async () => {
  render(<Form isSubmitting={false} onSubmit={jest.fn()} />)

  const button = screen.getByRole('button');
  fireEvent.click(button);

  await waitFor(() => {
    const elements = screen.queryAllByText(/this field is required/i);
    expect(elements).toHaveLength(3);
  });
});