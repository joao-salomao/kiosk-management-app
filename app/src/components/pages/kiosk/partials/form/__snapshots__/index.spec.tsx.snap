// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when is submitting the form renders correctly 1`] = `
<div
  className="flex flex-col justify-center"
>
  <div
    className="mx-auto shadow-lg rounded p-5"
    style={
      Object {
        "width": 900,
      }
    }
  >
    <p
      className="text-lg font-medium text-gray-900 mb-2"
    />
    <form
      onSubmit={[Function]}
    >
      <div
        className="mb-6"
      >
        <label
          className="block mb-2 text-sm font-medium text-gray-900"
          htmlFor="serialKey"
          id="serialKeyLabel"
        >
          Serial Key:
        </label>
        <input
          aria-labelledby="serialKeyLabel"
          className="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5"
          name="serialKey"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Ex: 123-NZD-123"
          type="text"
        />
      </div>
      <div
        className="mb-6"
      >
        <label
          className="block mb-2 text-sm font-medium text-gray-900"
          htmlFor="description"
          id="descriptionLabel"
        >
          Description:
        </label>
        <textarea
          aria-labelledby="descriptionLabel"
          className="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5"
          cols={33}
          name="description"
          onBlur={[Function]}
          onChange={[Function]}
          rows={2}
        />
      </div>
      <div
        className="mb-6"
      >
        <label
          className="block text-sm font-medium text-gray-900 mb-2"
          htmlFor="isKioskClosed"
          id="isKioskClosedLabel"
        >
          Is closed?
        </label>
        <label
          className="inline-flex relative items-center cursor-pointer"
        >
          <input
            aria-labelledby="isKioskClosedLabel"
            className="sr-only peer"
            name="isKioskClosed"
            onBlur={[Function]}
            onChange={[Function]}
            type="checkbox"
          />
          <div
            className="w-11 h-6 bg-gray-200 peer-focus:outline-none peer-focus:ring-4 peer-focus:ring-blue-300 rounded-full peer peer-checked:after:translate-x-full peer-checked:after:border-white after:content-[''] after:absolute after:top-[2px] after:left-[2px] after:bg-white after:border-gray-300 after:border after:rounded-full after:h-5 after:w-5 after:transition-all peer-checked:bg-blue-600"
          />
        </label>
      </div>
      <div
        className="mb-6"
      >
        <div
          className="flex"
        >
          <div
            className="mr-4"
          >
            <label
              className="block mb-2 text-sm font-medium text-gray-900"
              htmlFor="storeOpensAt"
              id="storeOpensAtLabel"
            >
              Store opens at:
            </label>
            <input
              aria-labelledby="storeOpensAtLabel"
              className="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5"
              name="storeOpensAt"
              onBlur={[Function]}
              onChange={[Function]}
              type="time"
            />
          </div>
          <div>
            <label
              className="block mb-2 text-sm font-medium text-gray-900"
              htmlFor="storeClosesAt"
              id="storeClosesAtLabel"
            >
              Store closes at:
            </label>
            <input
              aria-labelledby="storeClosesAtLabel"
              className="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5"
              name="storeClosesAt"
              onBlur={[Function]}
              onChange={[Function]}
              type="time"
            />
          </div>
        </div>
      </div>
      <button
        className="float-right text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 mr-2 focus:outline-none h-10 mt-auto"
        disabled={true}
        type="submit"
      >
        Saving
      </button>
    </form>
  </div>
</div>
`;
